### PR TITLE
Fix setting of rate_prototype in __init for ArrayPartition types

### LINKED ISSUE #1308: Vector of StaticArrays inside ArrayPartition breaks the type of rate_prototype
I have noticed, that if the state vector `u` is an `ArrayPartition` which contains `Vector` of `StaticArray`s, the computed type of the derivative vector `du` comes out wrong. I believe the problem resides in this line: https://github.com/SciML/OrdinaryDiffEq.jl/blob/69484227aa2b069d42f2c269339c84e34bbfa127/src/solve.jl#L184
Consider the following example:
```
using StaticArrays, RecursiveArrayTools

tType = Float64
u = ArrayPartition((zeros(5), zeros(SVector{3,Float64}, 5)))
typeof.(oneunit.(recursive_bottom_eltype.(u.x))./oneunit(tType))
```
gives you
>(Float64, Float64)

As a result, when `rate_prototype` is created, we get `Vector{Float64}` instead of `Vector{SVector{3,Float64}}` for the second part of `ArrayPartition`.

Since the line https://github.com/SciML/OrdinaryDiffEq.jl/blob/69484227aa2b069d42f2c269339c84e34bbfa127/src/solve.jl#L184 is supposed to correspond to the case without units, it actually looks completely out of place. The fix seems to be just to remove this line and calculate `rate_prototype` the same way it is done when `u` is not `ArrayPartition` in line https://github.com/SciML/OrdinaryDiffEq.jl/blob/69484227aa2b069d42f2c269339c84e34bbfa127/src/solve.jl#L182

### PR DESCRIPTION
**Pull request by jdeldre**
I removed the negation on the conditional `typeof(u) <: ArrayPartition` to enable use of `recursivecopy` to set `rate_prototype` for `ArrayPartition` type `u`.

Fixes #1308

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Looks like that's a real test failure.

File: src/solve.jl [hunk 1]
```diff
@@ -178,7 +178,7 @@ function DiffEqBase.__init(prob::Union{DiffEqBase.AbstractODEProblem,DiffEqBase.
   # dtmin is all abs => does not care about sign already.
 
   if !isdae && isinplace(prob) && typeof(u) <: AbstractArray && eltype(u) <: Number && uBottomEltypeNoUnits == uBottomEltype && tType == tTypeNoUnits # Could this be more efficient for other arrays?
-    if !(typeof(u) <: ArrayPartition)
+    if typeof(u) <: ArrayPartition
```
Comment (devmotion) [inline src/solve.jl:181]: This doesn't seem like a good idea. Before, everything apart from ArrayPartition ended up in this branch, and with this change only ArrayPartition would end up in this branch. It seems what you would like is rather to remove the other branch and always use `rate_prototype = recursivecopy(u)`.

File: src/solve.jl [hunk 2]
```diff
@@ -178,7 +178,7 @@ function DiffEqBase.__init(prob::Union{DiffEqBase.AbstractODEProblem,DiffEqBase.
...
       rate_prototype = recursivecopy(u)
     else
       rate_prototype = similar(u, typeof.(oneunit.(recursive_bottom_eltype.(u.x))./oneunit(tType))...)
```
Comment (devmotion) [inline src/solve.jl:184]: So maybe rather:
```suggestion
    rate_prototype = recursivecopy(u)
```
However, I didn't test it, so maybe it fixes the mentioned issue but breaks some other ArrayPartition cases...

File: src/solve.jl [hunk 3]
```diff
@@ -178,10 +178,7 @@ function DiffEqBase.__init(prob::Union{DiffEqBase.AbstractODEProblem,DiffEqBase.
   # dtmin is all abs => does not care about sign already.
 
   if !isdae && isinplace(prob) && typeof(u) <: AbstractArray && eltype(u) <: Number && uBottomEltypeNoUnits == uBottomEltype && tType == tTypeNoUnits # Could this be more efficient for other arrays?
-    if !(typeof(u) <: ArrayPartition)
-      rate_prototype = recursivecopy(u)
-    else
-      rate_prototype = similar(u, typeof.(oneunit.(recursive_bottom_eltype.(u.x))./oneunit(tType))...)
+    rate_prototype = recursivecopy(u)
     end
```
Comment (devmotion) [inline src/solve.jl:182]:
```suggestion
```

**Comment by devmotion** [comment]
It would be good to add a small example of the issue that is fixed here to the tests.

**Comment by devmotion** [comment]
It seems this change breaks other `ArrayPartition`s (see, e.g., https://github.com/SciML/OrdinaryDiffEq.jl/pull/1309/checks?check_run_id=1563518704#step:6:255), which is probably what motivates the current implementation.

**Comment by jdeldre** [comment]
Got ahead of myself. This failure is fixed by https://github.com/SciML/DiffEqBase.jl/pull/621. New release is imminent.

**Review by devmotion — APPROVED** [review]
LGTM. Thanks for the PR!

**Comment by jdeldre** [comment]
@ChrisRackauckas Do you think you could issue a new release? I've got a few packages that depend on this fix.

**Comment by ChrisRackauckas** [comment]
Done.

**Pull request merged**